http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/http2/Http2MultiplexedChannelPool.java

```diff
@@ -46,151 +58,333 @@
+        parentChannel.pipeline().addLast(new ReleaseOnExceptionHandler());
+
+        connections.add(multiplexedChannel);
+
+        if (closed.get()) {
+            // Whoops, we were closed while we were setting up. Make sure everything here is cleaned up properly.
+            failAndCloseParent(promise, parentChannel,
+                               new IOException("Connection pool was closed while creating a new stream."));
+            return;
         }
-        connections.remove(record);
+
+        promise.setSuccess(stream);
+    }
+
+    private Void failAndCloseParent(Promise<Channel> promise, Channel parentChannel, Throwable exception) {
+        promise.setFailure(exception);
+        closeAndReleaseParent(parentChannel);
+        return null;
+    }
+
+    /**
+     * Acquire a stream on a connection that has already been initialized. This will return false if the connection cannot have
+     * any more streams allocated, and true if the stream can be allocated.
+     *
+     * This will NEVER complete the provided future when the return value is false. This will ALWAYS complete the provided
+     * future when the return value is true.
+     */
+    private boolean acquireStreamOnInitializedConnection(MultiplexedChannelRecord channelRecord, Promise<Channel> promise) {
+        Promise<Channel> acquirePromise = channelRecord.getConnection().eventLoop().newPromise();
+
+        if (!channelRecord.acquireStream(acquirePromise)) {
+            return false;
+        }
+
+        acquirePromise.addListener(f -> {
+            try {
+                if (!acquirePromise.isSuccess()) {
+                    promise.setFailure(acquirePromise.cause());
+                    return;
+                }
+
+                Channel channel = acquirePromise.getNow();
+                channel.attr(ChannelAttributeKey.HTTP2_MULTIPLEXED_CHANNEL_POOL).set(this);
+                channel.attr(MULTIPLEXED_CHANNEL).set(channelRecord);
+                promise.setSuccess(channel);
+            } catch (Exception e) {
+                promise.setFailure(e);
+            }
+        });
+
+        return true;
     }
 
     @Override
     public Future<Void> release(Channel childChannel) {
-        return release(childChannel, new DefaultPromise<>(eventLoop));
+        return release(childChannel, childChannel.eventLoop().newPromise());
     }
 
     @Override
-    public Future<Void> release(Channel channel, Promise<Void> promise) {
-        doInEventLoop(eventLoop, () -> release0(channel, promise), promise);
-        return promise;
+    public Future<Void> release(Channel childChannel, Promise<Void> promise) {
+        if (childChannel.parent() == null) {
+            // This isn't a child channel. Oddly enough, this is "expected" and is handled properly by the
+            // BetterFixedChannelPool AS LONG AS we return an IllegalArgumentException via the promise.
```

Comment:
Because of this check: https://github.com/aws/aws-sdk-java-v2/blob/master/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/utils/BetterFixedChannelPool.java#L217-L220